apps/passport-server/src/services/generic-issuance/SemaphoreGroupProvider.ts

```diff
@@ -0,0 +1,309 @@
+      const group = this.groups.get(groupConfig.groupId);
+      if (!group) {
+        throw new Error(
+          `Could not find group ${groupConfig.groupId} on pipeline ${this.pipelineId}`
+        );
+      }
+
+      const { toAdd, toRemove } = this.calculateGroupChanges(
+        group,
+        semaphoreIds
+      );
+
+      if (toAdd.length > 0 || toRemove.length > 0) {
+        for (const newId of toAdd) {
+          group.addMember(newId);
+        }
+
+        for (const deletedId of toRemove) {
+          group.removeMember(group.indexOf(BigInt(deletedId)));
+        }
+
+        await this.semaphoreHistoryDB.add(
+          this.pipelineId,
+          groupConfig.groupId,
+          group.root.toString(),
+          JSON.stringify(serializeSemaphoreGroup(group, groupConfig.name))
+        );
+      }
+      span?.setAttribute("pipeline_id", this.pipelineId);
+      span?.setAttribute("emails_count", emails.length);
+      span?.setAttribute("consumers_count", consumers.length);
+      span?.setAttribute("group_id", groupConfig.groupId);
+      span?.setAttribute("added_count", toAdd.length);
+      span?.setAttribute("removed_count", toRemove.length);
+      span?.setAttribute("total_members_after_update", group.members.length);
+    });
+  }
+
+  /**
+   * Calculates the changes to the group compared to the latest set of
+   * members.
+   */
+  private calculateGroupChanges(
+    group: Group,
+    latestMembers: string[]
+  ): { toAdd: string[]; toRemove: string[] } {
+    const groupMembers = group.members
+      .filter((m) => m !== group.zeroValue)
+      .map((m) => m.toString());
+
+    const groupMemberSet = new Set(groupMembers);
+
+    const toAdd = latestMembers.filter((id) => !groupMemberSet.has(id));
+    const latestMemberSet = new Set(latestMembers);
+    const toRemove = groupMembers.filter(
+      (id) => !latestMemberSet.has(id.toString())
+    );
+
+    return {
+      toAdd,
+      toRemove
+    };
+  }
+
+  /**
+   * Given data on the current set of tickets (specifically event ID, product
+   * ID, and email), match these to configured Semaphore groups.
+   */
+  private groupEmailsFromTicketData(
+    ticketData: SemaphoreGroupTicketInfo[]
+  ): Map<SemaphoreGroupConfig, string[]> {
+    const groups = new Map<SemaphoreGroupConfig, string[]>();
+    for (const ticketInfo of ticketData) {
+      for (const groupConfig of this.groupConfigs) {
+        const groupMemberEmails = groups.get(groupConfig) ?? [];
+        for (const ticketSpec of groupConfig.membershipTickets) {
+          if (
+            ticketSpec.eventId === ticketInfo.eventId &&
+            (!ticketSpec.productId ||
+              ticketSpec.productId === ticketInfo.productId)
+          ) {
+            groupMemberEmails.push(ticketInfo.email);
+          }
+        }
+        groups.set(groupConfig, groupMemberEmails);
+      }
+    }
+    return groups;
+  }
+
+  public getSupportedGroups(): PipelineSemaphoreGroups {
```

Comment:
should this return `undefined` if `this.groupConfigs` is undefined?

Reply:
I don't have a strong feeling about it, but an empty array seems to have the same meaning.